### ISSUE TITLE
Illegal Instruction (core dumped)

### ISSUE THREAD

**Issue by superblaubeere27** [simd-lite/simd-json#200]
Important to note:
- According to `cat /proc/cpuinfo`, my cpu supports SSE4.2.
- I have the `.cargo/config` of this repository in place
- I am using rustc 1.53.0

After compiling the program, it crashes with the message `Illegal Instruction (core dumped)`.

How can I fix this?

**Comment by Licenser** [comment]
There are a few possibilities here, usually the issue means that the compilation was done with features that aren't supported by the CPU, I got a few ideas on that :)

1) Do you compile and run the code on the same machine? If yes then you can use `rustflags = ["-C", "target-cpu=native"]` in the `.cargo/config`
2) If you do not, you might have a wrong setting in the `.cargo/config` if the CPU you run it on has SSE4.2 the setting that **should** :tm: work is `rustflags = ["-C", "target-feature=+sse4.2"]`

If neither of that works could you share your `/proc/cpuinfo` and `.cargo/config`? :)

**Comment by superblaubeere27** [comment]
Thank you for your quick reply.

I compiled the program on the same machine I ran it on and I had the `rustflags = ["-C", "target-cpu=native"] ` in `.cargo/config`.

Unfortunately also the setting you suggested (`+sse4.2`) did not work and the program still crashes due to the CPU not recognizing an instruction.

Here are the requsted files:

### `.cargo/config` (Right now):
```TOML
[build]
rustflags = ["-C", "target-feature=+sse4.2"]
```
### `.cargo/config` (When making this issue):
```TOML
[build]
rustflags = ["-C", "target-feature=native"]
```
### `car /proc/cpuinfo`:
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 26
model name      : Intel(R) Xeon(R) CPU           E5530  @ 2.40GHz
stepping        : 5
microcode       : 0x1d
cpu MHz         : 2533.562
cache size      : 8192 KB
physical id     : 0
siblings        : 8
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 11
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm dca sse4_1 sse4_2 popcnt lahf_lm pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid dtherm ida flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit
bogomips        : 4800.43
clflush size    : 64
cache_alignment : 64
address sizes   : 40 bits physical, 48 bits virtual
power management:

...
```

**Comment by Licenser** [comment]
This is incredibly odd o.O is the project you're working on open source? Is that a physical machine or a VM?

**Comment by superblaubeere27** [comment]
The project is not open source, I can try to create a minimal reproducable example though.

I am using a real physical machine running Ubuntu 20.04.

**Comment by superblaubeere27** [comment]
I found out why this happens:

The instruction `pclmulqdq` is generated in [`sse4.2/stage1.rs:61`](https://github.com/simd-lite/simd-json/blob/main/src/sse42/stage1.rs#L61). According to intel's [reference](https://web.archive.org/web/20210704024613/https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_clmulepi64_si128&expand=682) this instruction is not part of SSE4.2 and more recent. It was introduced with later nehalem models, which my CPU is not part of.

This should be added to the documentation and simd-json should check if the CPU supports `PCLMULQDQ` instructions. Alternatively, the instruction could just be replaced with something else.

**Comment by Licenser** [comment]
darn, that's a good investigation! https://doc.rust-lang.org/core/arch/x86_64/fn._mm_clmulepi64_si128.html this is still odd to see however since I suspect it should fail when `pclmulqdq` isn't set.

Out of curiosity what does `rustc --print target-features` say with `cpu-target=native`?

**Comment by superblaubeere27** [comment]
I executed `rustc --print target-features` in the project folder. It just printed every feature that it's code generation supports.´

**Comment by Licenser** [comment]
Sorry yes, that was the wrong one :) I don't use that commands often enough `rustc --print cfg -C target-cpu=native` was what I had been looking for

**Comment by superblaubeere27** [comment]
Prints exactly what you would expect:
```
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="gnu"
target_family="unix"
target_feature="fxsr"
target_feature="popcnt"
target_feature="sse"
target_feature="sse2"
target_feature="sse3"
target_feature="sse4.1"
target_feature="sse4.2"
target_feature="ssse3"
target_os="linux"
target_pointer_width="64"
target_vendor="unknown"
unix
```

On my *Desktop-PC* (which it works on) it prints this:
```
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="msvc"
target_family="windows"
target_feature="aes"
target_feature="avx"
target_feature="avx2"
target_feature="bmi1"
target_feature="bmi2"
target_feature="fma"
target_feature="fxsr"
target_feature="lzcnt"
target_feature="pclmulqdq"
target_feature="popcnt"
target_feature="rdrand"
target_feature="rdseed"
target_feature="sse"
target_feature="sse2"
target_feature="sse3"
target_feature="sse4.1"
target_feature="sse4.2"
target_feature="ssse3"
target_feature="xsave"
target_feature="xsavec"
target_feature="xsaveopt"
target_feature="xsaves"
target_os="windows"
target_pointer_width="64"
target_vendor="pc"
windows
```

**Comment by Licenser** [comment]
I'm genuenly surprised it even compiles, the docs claim `This is supported on x86-64 and target feature pclmulqdq only` which, as you point out, isn't the case on your target machine. I would have expected rust to say "No you can't compile `_mm_clmulepi64_si128` isn't a function

**Comment by Licenser** [comment]
I've released your PR  as 0.4.7

**Comment by superblaubeere27** [comment]
👍🏻
